File: 111-minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root) return 0;
        if (!root->left) return 1 + minDepth(root->right);
        if (!root->right) return 1 + minDepth(root->left);

        return 1 + min(minDepth(root->left), minDepth(root->right));
    }
};

// #include <queue>
// class Solution {
// public:
//     int minDepth(TreeNode* root) {
//         if (!root) return 0;

//         std::queue<std::pair<TreeNode*, int>> q;
//         q.push({root, 1});

//         while (!q.empty()) {
//             TreeNode* node = q.front().first;
//             int depth = q.front().second;
//             q.pop();

//             if (!node->left && !node->right) {
//                 return depth; // First leaf reached
//             }

//             if (node->left) q.push({node->left, depth + 1});
//             if (node->right) q.push({node->right, depth + 1});
//         }

//         return 0; // Should never reach here
//     }
// };

File: 111-minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def minDepth(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        if not root:
            return 0

        q = deque([root])
        depth = 1
        while q:
            current_size = len(q)
            for i in range(current_size):
                node = q.popleft()
                if not node.left and not node.right:
                    return depth

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            depth+=1
        return depth